jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PoolingPathResource.java

```diff
@@ -0,0 +1,265 @@
+    {
+        assert POOL_LOCK.isHeldByCurrentThread();
+
+        Metadata metadata = POOL.get(fileSystem);
+        if (metadata == null)
+        {
+            LOG.debug("Pooling new FS {}", fileSystem);
+            metadata = new Metadata(uri);
+            POOL.put(fileSystem, metadata);
+        }
+        else
+        {
+            int count = metadata.counter.incrementAndGet();
+            LOG.debug("Incremented ref counter to {} for FS {}", count, fileSystem);
+        }
+        return metadata;
+    }
+
+    private static void release(FileSystem fileSystem)
+    {
+        assert POOL_LOCK.isHeldByCurrentThread();
+
+        Metadata metadata = POOL.get(fileSystem);
+        int count = metadata.counter.decrementAndGet();
+        if (count == 0)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("Ref counter reached 0, closing pooled FS {}", fileSystem);
+            IO.close(fileSystem);
+            POOL.remove(fileSystem);
+        }
+        else
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("Decremented ref counter to {} for FS {}", count, fileSystem);
+        }
+    }
+
+    private static class Metadata
+    {
+        private final AtomicInteger counter;
+        private final FileTime lastModifiedTime;
+        private final long size;
+        private final Path path;
+        private final Mount mount;
+
+        private Metadata(URI uri)
+        {
+            Path path = uriToPath(uri);
+
+            long size = -1L;
+            FileTime lastModifiedTime = null;
+            if (path != null)
+            {
+                try
+                {
+                    size = Files.size(path);
+                    lastModifiedTime = Files.getLastModifiedTime(path);
+                }
+                catch (IOException ioe)
+                {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("Cannot read size or last modified time from {} backing filesystem at {}", path, uri);
+                }
+            }
+            else
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Filesystem at {} is not backed by a file", uri);
+            }
+            this.counter = new AtomicInteger(1);
+            this.path = path;
+            this.size = size;
+            this.lastModifiedTime = lastModifiedTime;
+            this.mount = new Mount(uri);
+        }
+
+        private static Path uriToPath(URI uri)
+        {
+            URI rawUri = containerUri(uri);
+            return rawUri == null ? null : Paths.get(rawUri).toAbsolutePath();
+        }
+    }
+
+    public static class Mount implements Closeable
+    {
+        private final URI uri;
+
+        private Mount(URI uri)
+        {
+            this.uri = uri;
+        }
+
+        public Resource newResource() throws IOException
+        {
+            return Resource.newResource(uri);
+        }
```

Comment:
I'm not sure we need to make a new resource for every call to this?
Wouldn't it be better/clearer to have:
```suggestion
        final Resource root = Resource.newResource(uri);
        
        // ...
        
        public Resource root()
        {
            return root;
        }
```

Reply:
IMHO It's not something super desirable to be able to call `newResource()` many times, but it would be a serious annoyance to have to deal with the `IOException` in the constructor.

Reply:
I think it is better to deal with the Exception when the resource is created rather than when it is first used.

Reply:
Ok, done.